### PR TITLE
buck: 2019.10.17.01 -> 2021.01.12.01

### LINKED ISSUE #119224: Buck versions don't match
Buck as installed via nix reports with an older release version than the one listed in its nixpkgs default.nix. My guess is that `buck --version` is probably the correct version. The issue here is that IIRC buck

```
$ nix-shell -p buck                                                                                                                                   :(

[nix-shell:~/goals/fumoko]$ buck --version
buck version v2018.06.25.01.

[nix-shell:~/goals/fumoko]$ which buck
/nix/store/2v2lqldfkzgzvbq1rjzhph2hsk11hy6n-buck-2019.10.17.01/bin/buck
```

**Expected behavior**
`buck --version`'s output should match the version rerefenced in the build file.

**Notify maintainers**

@jgertm @marsam

### PR DESCRIPTION
**Pull request by jonringer**
###### Motivation for this change
closes #119224

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/build-managers/buck/default.nix [hunk 1]
```diff
@@ -17,20 +17,21 @@ stdenv.mkDerivation rec {
...
     PYTHONDONTWRITEBYTECODE=true ./bin/buck build -c buck.release_version=${version} buck
   '';
 
   installPhase = ''
-    install -D -m755 buck-out/gen/programs/buck.pex $out/bin/buck
+    install -D -m755 buck-out/gen/*/programs/buck.pex $out/bin/buck
     wrapProgram $out/bin/buck \
-      --prefix PYTHONPATH : $PYTHONPATH \
-      --prefix PATH : "${lib.makeBinPath [jdk watchman]}"
+      --prefix PATH : "${lib.makeBinPath [ jdk11 watchman python3 ]}"
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/buck/default.nix:34]: Those shouldn't be in nativeBuildInputs, right?

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 119247` at 8da4120a run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>buck</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/tools/build-managers/buck/default.nix:14:15:
```
   |
14 |   patches = [ ./pex-mtime.patch ];
   |               ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/tools/build-managers/buck/default.nix:31:3:
```
   |
31 |   installPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/tools/build-managers/buck/default.nix:22:3:
```
   |
22 |   buildPhase = ''
   |   ^
```
</li>
  </ul>
</details>

File: pkgs/development/tools/build-managers/buck/default.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/development/tools/build-managers/buck/default.nix:34]: They need to be in both, we are building buck with them present. Then buck expects them at runtime

**Comment by jonringer** [comment]
cc @HolgerPeters

File: pkgs/development/tools/build-managers/buck/default.nix [hunk 2]
```diff
@@ -17,20 +17,21 @@ stdenv.mkDerivation rec {
     grep -l -r '/bin/bash' --null | xargs -0 sed -i -e "s!/bin/bash!${bash}/bin/bash!g"
   '';
 
-  buildInputs = [ jdk ant python2 watchman python2Packages.pywatchman ];
-  nativeBuildInputs = [ makeWrapper ];
+  nativeBuildInputs = [ makeWrapper python3 jdk11 ant watchman ];
```
Comment (HolgerPeters) [inline pkgs/development/tools/build-managers/buck/default.nix:20]: I think this references the wrong Jak
```
% buck
Error: A JNI error has occurred, please check your installation and try again
Exception in thread "main" java.lang.Error: java.lang.NullPointerException
	at java.base/java.nio.file.FileSystems$DefaultFileSystemHolder.getDefaultProvider(FileSystems.java:133)
	< snip >
	... 34 more
You're using Java 11, but Buck requires Java 8.
```

Comment (HolgerPeters) [reply pkgs/development/tools/build-managers/buck/default.nix:20]: Created #120557 for that issue

**Pull request merged**